src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
         {
             AssertIsForeground();
 
-            using (Dispatcher.CurrentDispatcher.DisableProcessing())
+            // If the solution is closing we shouldn't do anything, because all of our state is
+            // in the process of going away. This can happen if we receive notification that a document has
+            // opened in the middle of the solution close operation.
+            if (_solutionIsClosing)
+            {
+                return;
+            }
+
+            // We need to push these projects and any project dependencies we already know about. Therefore, compute the
+            // transitive closure of the projects that haven't already been pushed, keeping them in appropriate order.
+            var visited = new HashSet<AbstractProject>();
+            var inOrderToPush = new List<AbstractProject>();
+
+            void addToInOrderToPush(AbstractProject project)
             {
-                foreach (var hostState in _workspaceHosts)
+                Contract.ThrowIfFalse(ContainsProject(project));
+
+                // Bail out if any of the following is true:
+                //  1. We have already started pushing changes for this project OR
+                //  2. We have already visited this project in a prior recursive call                
+                if (_pushedProjects.Contains(project) || !visited.Add(project))
+                {
+                    return;
+                }
+
+                foreach (var projectReference in project.GetCurrentProjectReferences())
```

Comment:
```GetCurrentProjectReferences```?  As opposed to what?  Previous project references?  Future references? (I'm not being snarky, i'm just tryign to figure out if there are special semantics here that i should be thinking about).

Reply:
This code is being moved, and deleted in another refactoring or two.